weblate_web/crm/views.py

```diff
@@ -336,3 +329,406 @@ def render_to_response(self, context, **response_kwargs):
+            svg_parts.append(
+                f'<text x="{label_x}" y="{label_y}" text-anchor="middle" '
+                f'font-size="{font_size}" fill="#666">{bar_label}</text>'
+            )
+
+        svg_parts.append("</svg>")
+        return "".join(svg_parts)
+
+    def _get_invoices_and_totals(
+        self, year: int, month: int | None = None
+    ) -> tuple[list, dict]:
+        """Fetch invoices and pre-calculate totals (shared helper)."""
+        query = Invoice.objects.filter(kind=InvoiceKind.INVOICE, issue_date__year=year)
+        if month:
+            query = query.filter(issue_date__month=month)
+
+        invoices = list(query.prefetch_related("invoiceitem_set"))
+        invoice_totals = {inv.pk: inv.total_amount_no_vat for inv in invoices}
+
+        return invoices, invoice_totals
+
+    def get_income_data(
+        self, year: int, month: int | None = None
+    ) -> dict[str, Decimal]:
+        """Get income data aggregated by category."""
+        invoices, invoice_totals = self._get_invoices_and_totals(year, month)
+
+        # Aggregate by category manually since total_amount_no_vat is a property
+        category_data = {}
+        for category in InvoiceCategory:
+            total = sum(
+                (
+                    invoice_totals[inv.pk]
+                    for inv in invoices
+                    if inv.category == category.value
+                ),
+                start=Decimal(0),
+            )
+            category_data[category.label] = total
+
+        return category_data
+
+    def get_monthly_data(self, year: int) -> tuple[dict[str, Decimal], list]:
+        """Get monthly income data for the year."""
+        invoices, invoice_totals = self._get_invoices_and_totals(year)
+
+        # Group by month in Python
+        monthly_totals = {}
+        for month in range(1, 13):
+            total = sum(
+                (
+                    invoice_totals[inv.pk]
+                    for inv in invoices
+                    if inv.issue_date.month == month
+                ),
+                start=Decimal(0),
+            )
+            monthly_totals[f"{month:02d}"] = total
+
+        return monthly_totals, invoices
+
+    def get_daily_data(self, year: int, month: int) -> tuple[dict[str, Decimal], list]:
+        """Get daily income data for a specific month."""
+        invoices, invoice_totals = self._get_invoices_and_totals(year, month)
+
+        # Get number of days in month
+        num_days = calendar.monthrange(year, month)[1]
+
+        # Group by day
+        daily_totals = {}
+        for day in range(1, num_days + 1):
+            total = sum(
+                (
+                    invoice_totals[inv.pk]
+                    for inv in invoices
+                    if inv.issue_date.day == day
+                ),
+                start=Decimal(0),
+            )
+            daily_totals[str(day)] = total
+
+        return daily_totals, invoices
```

Comment:
All the `get_*_data` differ only in the filter. Share the code and accept the filter as parameter that could be passed as lambda. Please remember proper type annotations.